lib/model/folder.go

```diff
@@ -51,3 +62,105 @@ func (f *folder) scanSubdirsIfHealthy(subDirs []string) error {
 	}
 	return nil
 }
+
+// A function to provide the ability to validate and modify local changes,
+// before they are committed to the database
+// Default behavior is to apply the changes as-is, overwrite function as needed
+func (f *folder) validateAndUpdateLocalChanges(fs []protocol.FileInfo) []protocol.FileInfo {
+	// update the database
+	f.model.updateLocals(f.folderID, fs)
+
+	return fs
+}
+
+// moveforconflict renames a file to deal with sync conflicts
+func (f *folder) moveForConflict(name string, maxConflicts int) error {
+	if strings.Contains(filepath.Base(name), ".sync-conflict-") {
+		l.Infoln("Conflict for", name, "which is already a conflict copy; not copying again.")
+		if err := os.Remove(name); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
+
+	if maxConflicts == 0 {
+		if err := os.Remove(name); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+		return nil
+	}
+
+	ext := filepath.Ext(name)
+	withoutExt := name[:len(name)-len(ext)]
+	newName := withoutExt + time.Now().Format(".sync-conflict-20060102-150405") + ext
+	err := os.Rename(name, newName)
+	if os.IsNotExist(err) {
+		// We were supposed to move a file away but it does not exist. Either
+		// the user has already moved it away, or the conflict was between a
+		// remote modification and a local delete. In either way it does not
+		// matter, go ahead as if the move succeeded.
+		err = nil
+	}
+	if maxConflicts > -1 {
+		matches, gerr := osutil.Glob(withoutExt + ".sync-conflict-????????-??????" + ext)
+		if gerr == nil && len(matches) > maxConflicts {
+			sort.Sort(sort.Reverse(sort.StringSlice(matches)))
+			for _, match := range matches[maxConflicts:] {
+				gerr = os.Remove(match)
+				if gerr != nil {
+					l.Debugln("removing extra conflict", gerr)
+				}
+			}
+		} else if gerr != nil {
+			l.Debugln("globbing for conflicts", gerr)
+		}
+	}
+	return err
+}
+
+// deletedir attempts to delete the given directory
+func (f *folder) deleteDir(path string, file protocol.FileInfo, matcher *ignore.Matcher) error {
+	realName, err := rootedJoinedPath(path, file.Name)
+	if err != nil {
+		return err
+	}
+
+	// Delete any temporary files lying around in the directory
+	dir, _ := os.Open(realName)
+	if dir != nil {
+		files, _ := dir.Readdirnames(-1)
+		for _, dirFile := range files {
+			fullDirFile := filepath.Join(file.Name, dirFile)
+			if defTempNamer.IsTemporary(dirFile) || (matcher != nil && matcher.Match(fullDirFile).IsDeletable()) {
+				os.RemoveAll(filepath.Join(path, fullDirFile))
+			}
+		}
+		dir.Close()
+	}
+
+	return osutil.InWritableDir(os.Remove, realName)
+}
+
+// deletefile attempts to delete the given file
+// Takes sync conflict and versioning settings into account
+func (f *folder) deleteFile(path string, file protocol.FileInfo, ver versioner.Versioner, maxConflicts int) error {
+	var err error
+
+	realName, err := rootedJoinedPath(path, file.Name)
+	if err != nil {
+		return err
+	}
+
+	if maxConflicts > 0 {
```

Comment:
Uh, this is incorrect. The original code checks whether there is a conflict between the update and the index version.